### DISCUSSION TITLE
Optional automatic same origin iframes Highlight support using Snow ❄️

### DISCUSSION THREAD

**Discussion by weizman** [highlight/highlight#4401] [Ideas]
### Snow ❄️ 

In [MetaMask 🦊 ](https://github.com/MetaMask) we are developing a JS security shim called [Snow ❄️](https://github.com/lavamoat/snow).

Snow exposes the simplest API to hook a callback to the event of new realms creation.
In other words, you can provide Snow a callback, and it will be invoked with the creation of every new iframe in the page:

```javascript
SNOW(win => console.log('new iframe detected:', win));
```

### Proposal

I think Snow integration into Highlight can be useful, so for example if you pass the entire Highlight clientside logic through Snow as a callback, and then change the [XHR override](https://github.com/highlight/highlight/blob/main/sdk/client/src/listeners/network-listener/utils/xhr-listener.ts#L30) from `const XHR = XMLHttpRequest.prototype` to `const XHR = win.XMLHttpRequest.prototype` (and so on) then Hightlight's logic of monitoring **will apply not only to the top main realm but to inner iframes as well automatically with zero effort.**

This will allow you to enhance your visibility into iframes and other same origin realms in a page rather than only the main realm of the app.

Snow supports all major browsers and all major websites.

### Discussion

I'd like your opinion on this, would be cool to see Snow being relevant somehow to this project (Snow is licensed as MIT)

**Comment by jay-khatri** [comment]
Hey @weizman, this seems very cool. And your example makes sense; we do support nested iframes in highlight.io. My initial reaction, however, is that since most of our browser-level instrumentation happens in our fork of https://github.com/rrweb-io/rrweb, SNOW may be a better fit there. Curious what @Vadman97 thinks.

**Reply by weizman** [reply]
Thanks man! May i ask, how exactly do you support same origin realms in highlight? From my long term research on snow I learned how difficult this task is, so I'm curious. 
Do you mean that the user must implement highlight in their iframes manually? Because if so, snow will take that responsibility off of them!

**Reply by Vadman97** [reply]
Same origin recording of the DOM in the iframe is implemented in https://github.com/rrweb-io/rrweb, the OSS library we use.
Network/console recording is implemented by us, and we currently do not support network/console recording in same-origin iframes.

**Comment by Vadman97** [comment]
hey @weizman , thanks for sharing! rrweb would be a better fit overall for most of the recording logic, but we actually have a [request](https://github.com/highlight/highlight/issues/3934) to support console/network recording via monkey patch in iframes which we do ourselves, where this may be a fit. i'll take a look at your project when we get to this!

**Reply by weizman** [reply]
Highly recommend snow, it really focuses on monkey patching iframes for you and not only iframes but same origin realms in general. Will take a lot of load off of you. Can't wait to hear your thoughts on it.

**Reply by Vadman97** [reply]
Nice! Will take a look in the coming weeks when we prioritize network/console support for iframes.
